### DISCUSSION TITLE
Problems when installing MOOSE offline

### DISCUSSION THREAD

**Discussion by Ethan-xj** [idaholab/moose#28535] [Q&A Getting Started] [answered]
### Check these boxes if you have followed the posting rules.

- [X] I have consulted the [Posting Guidelines](https://github.com/idaholab/moose/discussions/18270).
- [X] I have searched the [Discussions Forum](https://github.com/idaholab/moose/discussions) and [MOOSE Framework Troubleshooting](https://mooseframework.inl.gov/help/troubleshooting.html) and have not found what I was looking for
- [X] Q&A Getting Started is the most appropriate category for my question (trouble installing, beginner user, ...)

### Issue or question about MOOSE

I followed the official offline installation guide and successfully built the PETSc library, but encountered issues when building the Libmesh library. I'm wondering if I might not have correctly understood the guide. Do I also need to manually download the dependencies required by Libmesh? I only downloaded the dependencies needed for PETSc using the ./scripts/update_and_rebuild_petsc.sh --CC=$CC --CXX=$CXX --FC=$FC --with-packages-download-dir=~/projects/downloads command and then directly ran the ./update_and_rebuild_libmesh.sh --skip-submodule-update command.

Ethan

**Comment by GiudGiud** [comment]
Hello

When you download moose before going offline, this part `git submodule foreach --recursive git submodule update --init` should have taken care of downloading all the libmesh dependencies

**Reply by Ethan-xj** [reply]
OK. I previously encountered network issues and didn't succeed, but now it's been stuck at `Cloning into '/home/ethan/projects/moose/large_media'...` for a long time. Is this normal?

**Reply by GiudGiud** [reply]
that one is a little slow it s a big folder
you also dont need it unless you are building documentation

**Reply by Ethan-xj** [reply]
@GiudGiud OK. I've already downloaded the packages and got the offline.tar.gz
But when I build the libmesh, I got the error:
```
checking for the flags needed to use pthreads... -pthread
configure: /vexcl/vexcl.hpp not found
configure: /include/vexcl/vexcl.hpp not found
checking whether we are using the Microsoft C++ compiler... no
checking for CL/cl.h... no
checking for OpenCL/cl.h... no
checking for windows.h... no
checking for OpenCL library... no
checking for boost/system/error_code.hpp... yes
checking for the Boost system library... no
configure: error: cannot find flags to link with the Boost system library (libboost-system)
configure: error: ../../../contrib/metaphysicl/configure failed for contrib/metaphysicl
Running make -j 6...
make: *** No targets specified and no makefile found.  Stop.
```
The config is here: 
[config.log](https://github.com/user-attachments/files/16901722/config.log)

And when I run the diagnostics.sh, it goes like: 
```
################################################################################################
Python Sanity Checks

/usr/bin/env python3 --version; (reporting as: Python 3.11.1) matches
which python3 python;

/home/hpc4123303047/ZXY/MOOSE/packages/Python-3.11.1/build/bin/python3 --version; == Python 3.11.1
Python 2.7.5
FAIL: Python 2.7.5
/usr/bin/python --version (reporting as )) != Python 3.11.1

This will likely result in the TestHarness failing. Or WASP/HIT parsing errors.
################################################################################################
Python Modules (TestHarness, run-ability)

OK


checks FAILED
```
I'm not sure if the libmesh fail is because of the python version.
I can't access the /usr/bin, so how can I tell MOOSE to use the python 3.11.1 to build libmesh?

Ethan

**Reply by GiudGiud** [reply]
you could alias python to python3

if you go in libmesh/contrib/metaphyscl what do you see?
is there a config.log there? because I dont see much in the libmesh one
and can you configure?

**Reply by Ethan-xj** [reply]
Yes I've used alias like: alias python='/home/hpc4123303047/ZXY/MOOSE/packages/Python-3.11.1/build/bin/python3'
So, I don't need to refer to the Python check results in the shell script, right?
And I can configure in libmesh/contrib/metaphyscl:
![Quicker_20240906_095852](https://github.com/user-attachments/assets/2758a8a7-202d-4bd2-a3fd-4da392e302eb)
here is the config.log:
[config.log](https://github.com/user-attachments/files/16901873/config.log)

**Reply by GiudGiud** [reply]
it exits with code 1 so it failed to configure in metaphyscl
the error is
```
configure:35683: error: cannot find flags to link with the Boost system library (libboost-system)
```
I wonder if there is an option to get boost from libmesh or petsc. Your system boost seems to not work

**Reply by Ethan-xj** [reply]
[hpc4123303047@ln08 metaphysicl]$ yum list installed | grep boost
boost.x86_64                           1.53.0-27.el7                   @anaconda
boost-atomic.x86_64                    1.53.0-27.el7                   @anaconda
boost-chrono.x86_64                    1.53.0-27.el7                   @anaconda
boost-context.x86_64                   1.53.0-27.el7                   @anaconda
boost-date-time.x86_64                 1.53.0-27.el7                   @anaconda
boost-devel.x86_64                     1.53.0-27.el7                   @anaconda
boost-filesystem.x86_64                1.53.0-27.el7                   @anaconda
boost-graph.x86_64                     1.53.0-27.el7                   @anaconda
boost-iostreams.x86_64                 1.53.0-27.el7                   @anaconda
boost-locale.x86_64                    1.53.0-27.el7                   @anaconda
boost-math.x86_64                      1.53.0-27.el7                   @anaconda
boost-program-options.x86_64           1.53.0-27.el7                   @anaconda
boost-python.x86_64                    1.53.0-27.el7                   @anaconda
boost-random.x86_64                    1.53.0-27.el7                   @anaconda
boost-regex.x86_64                     1.53.0-27.el7                   @anaconda
boost-serialization.x86_64             1.53.0-27.el7                   @anaconda
boost-signals.x86_64                   1.53.0-27.el7                   @anaconda
boost-system.x86_64                    1.53.0-27.el7                   @anaconda
boost-test.x86_64                      1.53.0-27.el7                   @anaconda
boost-thread.x86_64                    1.53.0-27.el7                   @anaconda
boost-timer.x86_64                     1.53.0-27.el7                   @anaconda
boost-wave.x86_64                      1.53.0-27.el7                   @anaconda
ibus-typing-booster.noarch             1.2.3-4.el7                     @anaconda
My system has the Boost library installed, but I'm still encountering issues. How should I resolve this?

**Reply by Ethan-xj** [reply]
Perhaps I could try downloading a fresh copy of the boost library?

**Reply by GiudGiud** [reply]
That sounds like a good idea. In libmesh you can change the configure call to use a different boost.
You'll have to add those arguments to moose/scripts/configre_libmesh.sh if you go that route
```
  --with-boost[=ARG]      use Boost library from a standard location
  --with-boost-libdir=LIB_DIR
                          Force given directory for boost libraries. Note that
                          fails and you know exactly where your boost
```

**Reply by Ethan-xj** [reply]
@GiudGiud Hello Giud. I'm still stuck in this problem. I've installed the boost_1_85_0 from source. And in the bashrc file I set:
```
export CPLUS_INCLUDE_PATH=/home/hpc4123303047/ZXY/MOOSE/packages/boost_1_85_0/build/include:$CPLUS_INCLUDE_PATH
export LIBRARY_PATH=/home/hpc4123303047/ZXY/MOOSE/packages/boost_1_85_0/build/lib:$LIBRARY_PATH
export LD_LIBRARY_PATH=/home/hpc4123303047/ZXY/MOOSE/packages/boost_1_85_0/build/lib:$LD_LIBRARY_PATH
```
But the same errors appear when configuring:
```
checking for the Boost system library... no
configure: error: cannot find flags to link with the Boost system library (libboost-system)
```
Do you have any suggestions or good ideas on how to solve this issue? Thank you.

Ethan

**Reply by GiudGiud** [reply]
did you pass the commands ( --with-boost) when configuring libmesh?

**Reply by Ethan-xj** [reply]
Yes I run the command:  `./update_and_rebuild_libmesh.sh --skip-submodule-update --with-boost`
and the command: `./update_and_rebuild_libmesh.sh --skip-submodule-update --with-boost-libdir=/home/hpc4123303047/ZXY/MOOSE/packages/boost_1_85_0/build/lib`
But the error still exists:
```
checking for Boost headers version >= 1.47.0... yes
checking for Boost's header version... 1_85
checking for the toolset name used by Boost for mpicxx... configure: WARNING: could not figure out which toolset name to use for mpicxx

checking for the flags needed to use pthreads... -pthread
configure: /vexcl/vexcl.hpp not found
configure: /include/vexcl/vexcl.hpp not found
checking whether we are using the Microsoft C++ compiler... no
checking for CL/cl.h... no
checking for OpenCL/cl.h... no
checking for windows.h... no
checking for OpenCL library... no
checking for boost/system/error_code.hpp... yes
checking for the Boost system library... no
configure: error: cannot find flags to link with the Boost system library (libboost-system)
configure: error: ../../../contrib/metaphysicl/configure failed for contrib/metaphysicl
Running make -j 6...
make: *** No targets specified and no makefile found.  Stop.
```
So strange.

**Reply by Ethan-xj** [reply]
Should I add some debug statements in the `configure` script to determine where the issue is occurring?

**Reply by milljm** [reply]
My two cents: It looks like its finding multiple instances of Boost in different locations. Another observation, is that I think this is caused by OpenSSL somehow.

**Reply by Ethan-xj** [reply]
@milljm My openssl is like:
```
[hpc4123303047@ln08 scripts]$ openssl version
OpenSSL 1.1.1l  24 Aug 2021
```
I installed it recently.
And for the boost locations. It seems there is an older version of boost in the /usr directory, but I don't have access to that directory. Therefore, I installed a newer version in a different directory and specified my installation path during the configure process. Given that, it shouldn't be causing any issues.

**Reply by milljm** [reply]
Part of the issue I've noticed with libraries libMesh picks up in system locations, is that it now needs to include those with `-L/usr/lib` etc etc. And by doing so, it'll link with other libraries it may also need which are discovered in these locations. Usually this seems to be an issue with Boost and MPI (using system MPI grabs Boost also in this location by accident). But for your issue, it feels like something else is drawing its attention there.

Log example:
```pre
configure:35539: mpicxx -o conftest -g -O2  -Wall   -L/home/hpc4123303047/ZXY/MOOSE/packages/openssl-3.0.2/build/lib -L/lib64 conftest.o -lboost_system  >&5
/lib64/libboost_system.so: undefined reference to `operator delete(void*)@GLIBCXX_3.4'
/lib64/libboost_system.so: undefined reference to `__cxa_guard_release@CXXABI_1.3'
/lib64/libboost_system.so: undefined reference to `__cxa_guard_acquire@CXXABI_1.3'
/lib64/libboost_system.so: undefined reference to `vtable for __cxxabiv1::__class_type_info@CXXABI_1.3'
/lib64/libboost_system.so: undefined reference to `std::basic_string<char, std::char_traits<char>, std::allocator<char> >::~basic_string()@GLIBCXX_3.4'
```

Edit:
I made mention of OpenSSL, because I feel like _its_ the one doing it...

**Reply by Ethan-xj** [reply]
If the issue is with OpenSSL, and I have installed OpenSSL in the path /home/hpc4123303047/ZXY/MOOSE/packages/openssl-1.1.1l, what should I do next? Could the issue be related to the version of OpenSSL?

**Reply by milljm** [reply]
I don't think its a version of OpenSSL that is the issue, but perhaps by building your own, that in order to use it, it is requiring that we link to things in system locations (or something like that). I am not familiar with OpenSSL enough to know what it may be doing this for.

I only find it odd that in the config.log file, while I scroll through it quickly, I only see this link line pertaining to a custom OpenSSL version. Even though I would expect to see more (like all the other custom libraries you built manually).

**Reply by milljm** [reply]
Perhaps dropping this:
```pre
LDFLAGS='-L/home/hpc4123303047/ZXY/MOOSE/packages/openssl-3.0.2/build/lib'
```

**Comment by Ethan-xj** [comment]
My fault. Sry for that. I just realize that the config.log is changed.
Here is the new one:
[config.log](https://github.com/user-attachments/files/16932794/config.log)

**Reply by Ethan-xj** [reply]
@milljm I noticed that there's `--prefix=/usr/local` in it. But I can't access /usr/local. Could this be the source of the problem？

**Reply by milljm** [reply]
I was just going to report that, yes. Everything looked fine up until the contribs began to config.

Add a prefix argument:
```bash
--prefix=/home/hpc4123303047/ZXY/MOOSE/packages/libmesh
```
(just a guess as to where you want libmesh). Basically, I would place it right along side where ever you installed PETSc.

Where ever the location you pick, you'll want to then set `LIBMESH_DIR=` to point to this location, when you eventually get to the step of building MOOSE.

So all in all, in the end (if the above works):
```bash
export PETSC_DIR=/some/path/to/where/you/installed/petsc
export LIBMESH_DIR=/home/hpc4123303047/ZXY/MOOSE/packages/libmesh
cd ~/projects/moose/test
make
```

Edit: UGH... I always forget WASP :) You'll need to build WASP, and then set `WASP_DIR=` much in the same way above. And _then_ you can build MOOSE.

**Reply by Ethan-xj** [reply]
@milljm I'm afraid the issue still hasn't been resolved. I've set all the paths, including the prefix and specified the Boost path. However, it still throws errors. I ran the configure directly in the metaphysiclfolder(/home/hpc4123303047/projects/moose/libmesh/contrib/metaphysicl) and the config.log is here: 
[config.log](https://github.com/user-attachments/files/16957054/config.log)
Could you find any clues from it?

**Reply by Ethan-xj** [reply]
Could it be an issue with libboost_system-mt.so? I noticed that it appears multiple times in the config.log, but I don't see this file in my boost/build/lib folder.

**Reply by Ethan-xj** [reply, marked as answer]
The problem is solved. I use ./configure --help to check the rules of the metaphysic's configuration. I should use --with--boost=/root/to/boost. And it succeeds. Thank you all :)

**Reply by GiudGiud** [reply]
so you had to configure metaphysicl individually? The boost options did not propagate from libmesh?

**Reply by Ethan-xj** [reply]
No, I was just saying that I discovered the issue when compiling metaphysic separately. When I used the shell script `./update_and_rebuild_libmesh.sh --skip-submodule-update` with the --with-boost option, I successfully compiled libMesh, and I also successfully installed MOOSE and passed the tests.\
Thank you all :)

Ethan
